### ISSUE TITLE
Failed to use go-fuse mounts as lowerdirs of overlayfs since kernel 5.15

### ISSUE THREAD

**Issue by ktock** [hanwen/go-fuse#406]
Hi,

We're using go-fuse mounts as lowerdirs of overlayfs but it doesn't work since kernel 5.15.
go-fuse + overlayfs worked fine in the older version (5.14.17) though.

As shown in the following the debug log, overlayfs of kernel 5.15 calls IOCTL to lowerdirs when a write to the upperdir happens but go-fuse doesn't handle it and returns ENOSYS.
IOCTL didn't happen in 5.14.17.

- Ubuntu 20.04
- Linux 5.15.0-051500-generic
- go-fuse 3ab5d95a30ae (latest commit on master branch)

Overlayfs is mounted on `/mnt/snapshot` and the lowerdirs are go-fuse mounts.

```console
# touch /mnt/snapshot/bin/test
touch: cannot touch '/mnt/snapshot/bin/test': Function not implemented
```

mounts:

```console
# mount | grep -e " fuse" -e /mnt/snapshot
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/1/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/2/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/3/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/4/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/5/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/6/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/7/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/8/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/9/fs type fuse.rawBridge (rw,relatime,user_id=0,group_id=0,allow_other)
overlay on /mnt/snapshot type overlay (rw,relatime,lowerdir=/var/lib/containerd-stargz-grpc/snapshotter/snapshots/4/fs:/var/lib/containerd-stargz-grpc/snapshotter/snapshots/3/fs:/var/lib/containerd-stargz-grpc/snapshotter/snapshots/2/fs:/var/lib/containerd-stargz-grpc/snapshotter/snapshots/1/fs,upperdir=/var/lib/containerd-stargz-grpc/snapshotter/snapshots/10/fs,workdir=/var/lib/containerd-stargz-grpc/snapshotter/snapshots/10/work,xino=off)
```

Logs by go-fuse:

```
{"level":"debug","msg":"2021/11/11 02:20:52 rx 246: LOOKUP n1 [\"bin\"] 4b","time":"2021-11-11T02:20:52.475197956Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 246:     OK, {n2 g0 tE=1s tA=1s {M040755 SZ=0 L=2 0:0 B0*4096 i0:221 A 0.000000 M 1587603605.000000 C 0.000000}}","time":"2021-11-11T02:20:52.475306573Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 rx 488: LOOKUP n1 [\"bin\"] 4b","time":"2021-11-11T02:20:52.475405849Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 488:     OK, {n2 g0 tE=1s tA=1s {M040755 SZ=0 L=2 0:0 B0*4096 i0:461 A 0.000000 M 1587513600.000000 C 0.000000}}","time":"2021-11-11T02:20:52.475459166Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 rx 248: LISTXATTR n2 {sz 0} ","time":"2021-11-11T02:20:52.475598748Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 248:     OK","time":"2021-11-11T02:20:52.475633597Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 rx 250: OPENDIR n2 ","time":"2021-11-11T02:20:52.475668997Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 250:     OK, {Fh 1 }","time":"2021-11-11T02:20:52.475687058Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 rx 252: OPENDIR n2 ","time":"2021-11-11T02:20:52.475724596Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 252:     OK, {Fh 2 }","time":"2021-11-11T02:20:52.475785863Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 rx 254: IOCTL n2 *fuse._IoctlIn: \u0026{{72 39 254 2 {{0 0} 5482} 0} 2 16 2148034049 0 0 4} ","time":"2021-11-11T02:20:52.475822959Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 254:     38=function not implemented","time":"2021-11-11T02:20:52.475839601Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 rx 256: RELEASEDIR n2 {Fh 2   L0} ","time":"2021-11-11T02:20:52.475870731Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 256:     OK","time":"2021-11-11T02:20:52.475884140Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 rx 258: RELEASEDIR n2 {Fh 1   L0} ","time":"2021-11-11T02:20:52.476027462Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 258:     OK","time":"2021-11-11T02:20:52.476132451Z"}
```

**Comment by hanwen** [comment]
```
{"level":"debug","msg":"2021/11/11 02:20:52 rx 254: IOCTL n2 *fuse._IoctlIn: \u0026{{72 39 254 2 {{0 0} 5482} 0} 2 16 2148034049 0 0 4} ","time":"2021-11-11T02:20:52.475822959Z"}
{"level":"debug","msg":"2021/11/11 02:20:52 tx 254:     38=function not implemented","time":"2021-11-11T02:20:52.475839601Z"}
```

support for ioctls is very iffy (IIRC, I never attempted to implement anything.).

Can you discover what this ioctl is for?

**Comment by ktock** [comment]
@hanwen Sorry for the late reply.

Since commit https://github.com/torvalds/linux/commit/72db82115d2bdfbfba8b15a92d91872cfe1b40c6 of kernel, overlayfs tries to get fileattr flags from the lowerdir on copy up. This [results in calling IOCTL to FUSE](https://github.com/torvalds/linux/blob/72db82115d2bdfbfba8b15a92d91872cfe1b40c6/fs/fuse/ioctl.c#L419-L454).
In 5.15, overlayfs didn't allow it to return ENOSYS so we encountered this issue.

In the commit https://github.com/torvalds/linux/commit/5b0a414d06c3ed2097e32ef7944a4abb644b89bd (>= 5.16-rc1), mitigation is applied to this issue and ENOTTY or EINVAL is ignored during copy up.

I posted a PR https://github.com/hanwen/go-fuse/pull/408 to change the IOCTL implementation to return ENOTTY instead of ENOSYS.
This fixes the issue in 5.16 and later but not in 5.15.
To fully fix this, maybe we'll need to implement IOCTL.

**Comment by hanwen** [comment]
could you push to Gerrithub? See https://github.com/hanwen/go-fuse/blob/3ab5d95a30ae25019bf030a2bab076dba61c27da/CONTRIBUTING 

would it be possilbe to add a test?

**Comment by hanwen** [comment]
what does a libfuse FS do in this case? Would it not also fail?

**Comment by hanwen** [comment]
note to self: 

ioctl constants interpreted here: 
https://kernel.googlesource.com/pub/scm/linux/kernel/git/torvalds/linux/+/136057256686de39cc3a07c2e39ef6bc43003ff6/fs/ioctl.c#839
and
https://kernel.googlesource.com/pub/scm/linux/kernel/git/torvalds/linux/+/136057256686de39cc3a07c2e39ef6bc43003ff6/fs/ioctl.c#845

implementation https://kernel.googlesource.com/pub/scm/linux/kernel/git/torvalds/linux/+/136057256686de39cc3a07c2e39ef6bc43003ff6/fs/ioctl.c#739
and
https://kernel.googlesource.com/pub/scm/linux/kernel/git/torvalds/linux/+/136057256686de39cc3a07c2e39ef6bc43003ff6/fs/ioctl.c#708

there is a ENOIOCTLCMD errno value, 
https://kernel.googlesource.com/pub/scm/linux/kernel/git/torvalds/linux/+/136057256686de39cc3a07c2e39ef6bc43003ff6/include/linux/errno.h#17

which should be used to signal nonexistent ioctls ?

**Comment by ktock** [comment]
@hanwen 

> what does a libfuse FS do in this case? Would it not also fail?

For example, ntfs-3g seems to return EINVAL by default. So https://github.com/torvalds/linux/commit/5b0a414d06c3ed2097e32ef7944a4abb644b89bd allows lowerdir to return EINVAL.

https://github.com/tuxera/ntfs-3g/blob/a4a837025b6ac2b0c44c93e34e22535fe9e95b27/libntfs-3g/ioctl.c#L392-L419

> would it be possilbe to add a test?

Added a test but it's disabled by default because it requires root privilege to mount overlayfs and doesn't work on kernel 5.15 which doesn't contain https://github.com/torvalds/linux/commit/5b0a414d06c3ed2097e32ef7944a4abb644b89bd .

https://github.com/hanwen/go-fuse/pull/408/files#diff-654a5347b0564cde4e87530a517fe329bd243a5c8aabe7976043d79623ea4040

> could you push to Gerrithub?

Seems I need an additional github OAuth connection with gerrithub?
Couldn't we continue the review process on github?

**Comment by hanwen** [comment]
closing since this is fixed for >= 5.16
